In [ ]:
import torch
from transformers import BertTokenizer, BertModel
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd


In [ ]:
df = pd.read_csv('/content/merged_A61.csv')

In [ ]:
df=df[0:1000]

In [ ]:
# df = pd.read_csv('/content/Generated_Abstracts_A61B.csv')

In [ ]:
df.head()

patent_id                                    patent_abstract  \
0   11213381  A prosthetic capsular device configured to be ...   
1   11246717  A method and system for performing bone fusion...   
2   11234638  An oral appliance for the treatment of sleep d...   
3   11304855  Provided are buffered adhesive compositions co...   
4   11369552  An aqueous composition for removing nail polis...   

                                        patent_title  \
0  Prosthetic capsular devices, systems, and methods   
1                     Expandable intervertebral cage   
2  Intra-oral electroencephalography device and m...   
3  Buffered adhesive compositions for skin-adheri...   
4                        Aqueous nail polish remover   

                                          claim_text cpc_subclass  \
0  A prosthetic capsular system for insertion int...         A61F   
1  An intervertebral cage comprising: a first sup...         A61F   
2  An oral appliance for treatment of sleep apnea...         A61N   
3  An ostomy skin barrier comprising: a first pre...         A61L   
4  An aqueous composition for nail polish removal...         A61Q   

                                  generated_abstract  
0    The present invention relates to a prostheti...  
1    An intervertebral cage is described, compris...  
2    The present invention relates to an oral app...  
3    The present invention relates to an ostomy s...  
4    The present invention relates to an aqueous ...

In [ ]:
df.shape

(1000, 6)

In [ ]:
abstracts = df['patent_abstract'].tolist()
patent_ids = df['patent_id'].tolist()
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

def encode_texts(texts):
    encoded_inputs = tokenizer(texts, padding=True, truncation=True, return_tensors='pt', max_length=512)
    with torch.no_grad():
        outputs = model(**encoded_inputs)
    return outputs.last_hidden_state.mean(dim=1).cpu().numpy()
abstracts_embeddings = encode_texts(abstracts)

similarity_matrix = cosine_similarity(abstracts_embeddings)
similarities = similarity_matrix[0][1:]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

In [ ]:

closest_indices_20 = similarities.argsort()[-25:][::-1]
closest_patent_ids_20 = [patent_ids[i+1] for i in closest_indices_20]
l_30=[]
for i, patent_id in enumerate(closest_patent_ids_20):
    l_30.append(patent_id)
print(l_30)
print("=======================================")
closest_indices_10 = similarities.argsort()[-50:][::-1]
closest_patent_ids_10 = [patent_ids[i+1] for i in closest_indices_10]
l_50=[]
for i, patent_id in enumerate(closest_patent_ids_10):
  l_50.append(patent_id)
print(l_50)
print("=======================================")
closest_indices_5 = similarities.argsort()[-75:][::-1]
closest_patent_ids_5 = [patent_ids[i+1] for i in closest_indices_5]
l_75=[]
for i, patent_id in enumerate(closest_patent_ids_5):
    l_75.append(patent_id)
print(l_75)
print("=======================================")
closest_indices_3 = similarities.argsort()[-100:][::-1]
closest_patent_ids_3 = [patent_ids[i+1] for i in closest_indices_3]
l_100=[]
for i, patent_id in enumerate(closest_patent_ids_3):
    l_100.append(patent_id)
print(l_100)

[11234826, 11213431, 11229782, 11213676, 11224459, 11224459, 11213195, 11213195, 11253355, 11224720, 11311717, 11241318, 11213187, 11229772, 11260159, 11213675, 11273323, 11219708, 11219467, 11272956, 11266507, 11213331, 11235150, 11246715, 11219466]
[11234826, 11213431, 11229782, 11213676, 11224459, 11224459, 11213195, 11213195, 11253355, 11224720, 11311717, 11241318, 11213187, 11229772, 11260159, 11213675, 11273323, 11219708, 11219467, 11272956, 11266507, 11213331, 11235150, 11246715, 11219466, 11266520, 11213434, 11253144, 11253408, 11213317, 11259908, 11253379, 11229480, 11224499, 11246758, 11226501, 11234676, 11235116, 11246705, 11229513, 11272986, 11311454, 11291833, 11213323, 11229346, 11224428, 11233369, 11224451, 11241361, 11246582]
[11234826, 11213431, 11229782, 11213676, 11224459, 11224459, 11213195, 11213195, 11253355, 11224720, 11311717, 11241318, 11213187, 11229772, 11260159, 11213675, 11273323, 11219708, 11219467, 11272956, 11266507, 11213331, 11235150, 11246715, 1121946

Top 1 closest patent ID: 11234826
Top 2 closest patent ID: 11213431
Top 3 closest patent ID: 11229782
Top 4 closest patent ID: 11213676
Top 5 closest patent ID: 11224459
Top 6 closest patent ID: 11224459
Top 7 closest patent ID: 11213195
Top 8 closest patent ID: 11213195
Top 9 closest patent ID: 11253355
Top 10 closest patent ID: 11224720

In [ ]:
abstracts_g = df['generated_abstract'].tolist()
patent_ids = df['patent_id'].tolist()
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

def encode_texts(texts):
    encoded_inputs = tokenizer(texts, padding=True, truncation=True, return_tensors='pt', max_length=512)
    with torch.no_grad():
        outputs = model(**encoded_inputs)
    return outputs.last_hidden_state.mean(dim=1).cpu().numpy()

abstracts_embeddings_g = encode_texts(abstracts_g)

similarity_matrix_g = cosine_similarity(abstracts_embeddings_g)
similarities_g = similarity_matrix_g[0][1:]

In [ ]:
closest_indices_20 = similarities_g.argsort()[-25:][::-1]
closest_patent_ids_20 = [patent_ids[i+1] for i in closest_indices_20]
l_25_g=[]
for i, patent_id in enumerate(closest_patent_ids_20):
  l_25_g.append(patent_id)
print(l_25_g)
print("=======================================")

closest_indices_10 = similarities_g.argsort()[-50:][::-1]
closest_patent_ids_10 = [patent_ids[i+1] for i in closest_indices_10]
l_50_g=[]
for i, patent_id in enumerate(closest_patent_ids_10):
  l_50_g.append(patent_id)
print(l_50_g)
print("=======================================")
closest_indices_5 = similarities_g.argsort()[-75:][::-1]
closest_patent_ids_5 = [patent_ids[i+1] for i in closest_indices_5]
l_75_g=[]
for i, patent_id in enumerate(closest_patent_ids_5):
  l_75_g.append(patent_id)
print(l_75_g)
print("=======================================")
closest_indices_3 = similarities_g.argsort()[-100:][::-1]
closest_patent_ids_3 = [patent_ids[i+1] for i in closest_indices_3]
l_100_g=[]
for i, patent_id in enumerate(closest_patent_ids_3):
  l_100_g.append(patent_id)
print(l_100_g)

[11266521, 11272940, 11224459, 11219731, 11246715, 11224459, 11266490, 11219547, 11259908, 11253144, 11253386, 11241318, 11213297, 11241329, 11224451, 11219476, 11213331, 11234857, 11224458, 11213236, 11229552, 11213319, 11234827, 11229435, 11241305]
[11266521, 11272940, 11224459, 11219731, 11246715, 11224459, 11266490, 11219547, 11259908, 11253144, 11253386, 11241318, 11213297, 11241329, 11224451, 11219476, 11213331, 11234857, 11224458, 11213236, 11229552, 11213319, 11234827, 11229435, 11241305, 11260163, 11272956, 11253390, 11246758, 11260200, 11213388, 11246701, 11213195, 11213641, 11260159, 11219351, 11213675, 11298434, 11213195, 11230421, 11266784, 11241310, 11266518, 11241519, 11266418, 11219517, 11213236, 11272986, 11219462, 11213317]
[11266521, 11272940, 11224459, 11219731, 11246715, 11224459, 11266490, 11219547, 11259908, 11253144, 11253386, 11241318, 11213297, 11241329, 11224451, 11219476, 11213331, 11234857, 11224458, 11213236, 11229552, 11213319, 11234827, 11229435, 1124130

In [ ]:
A =set(l_30)
B = set(l_25_g)
# Intersaction and Union of two sets can also be done using & and | operators.
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11213331, 11224459, 11241318, 11246715}
AUB =  {11213187, 11219708, 11224451, 11259908, 11213319, 11234826, 11224459, 11213195, 11219467, 11219466, 11253386, 11224720, 11224458, 11234827, 11213331, 11219731, 11219476, 11253144, 11234857, 11229435, 11213236, 11266490, 11266507, 11229772, 11235150, 11229782, 11266521, 11241305, 11219547, 11311717, 11241318, 11253355, 11213676, 11213675, 11273323, 11272940, 11229552, 11213297, 11241329, 11213431, 11246715, 11272956, 11260159}
J(A,B) =  0.09302325581395349


In [ ]:
A=set(l_50)
B=set(l_50_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))


AnB =  {11224451, 11259908, 11213317, 11246758, 11241318, 11213195, 11224459, 11213675, 11213331, 11253144, 11272986, 11246715, 11272956, 11260159}
AUB =  {11260163, 11259908, 11213317, 11272956, 11219462, 11213319, 11234826, 11219467, 11219466, 11213323, 11234827, 11213331, 11219731, 11219476, 11272986, 11229480, 11260200, 11234857, 11235116, 11241519, 11219517, 11229513, 11213641, 11229772, 11213388, 11235150, 11229782, 11233369, 11241305, 11219547, 11311454, 11241310, 11311717, 11241318, 11253355, 11213676, 11213675, 11273323, 11224428, 11246701, 11246705, 11229552, 11241329, 11234676, 11266418, 11213431, 11213434, 11246715, 11298434, 11213187, 11253379, 11226501, 11224451, 11253386, 11224459, 11213195, 11224458, 11253390, 11224720, 11241361, 11219351, 11253144, 11253408, 11229346, 11246758, 11224499, 11213236, 11291833, 11266490, 11266507, 11230421, 11266518, 11266520, 11266521, 11266784, 11272940, 11213297, 11246582, 11229435, 11219708, 11260159}
J(A,B) =  0.1728395061728395


In [ ]:
A=set(l_75)
B=set(l_75_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11224451, 11259908, 11213317, 11224459, 11213195, 11219467, 11253390, 11213331, 11219351, 11253144, 11272986, 11229346, 11246758, 11235116, 11229771, 11266521, 11241318, 11213675, 11229552, 11241329, 11246705, 11213431, 11246715, 11272956, 11260159}
AUB =  {11260163, 11259908, 11213317, 11272956, 11219462, 11213319, 11234826, 11219467, 11219466, 11213323, 11266828, 11292173, 11234827, 11229452, 11213331, 11219731, 11219476, 11229462, 11272986, 11229476, 11229480, 11246632, 11260200, 11234857, 11235116, 11219756, 11241519, 11259953, 11213362, 11241267, 11235122, 11273015, 11235128, 11219517, 11259969, 11305028, 11229513, 11213641, 11229771, 11229772, 11213388, 11235150, 11246674, 11229523, 11229780, 11229782, 11233369, 11241305, 11219547, 11235164, 11311454, 11224415, 11241310, 11311717, 11241318, 11246953, 11253355, 11213676, 11213675, 11273323, 11224428, 11229552, 11246705, 11241329, 11246701, 11234676, 11266418, 11271281, 11213431, 11213173, 11213434, 11246715, 11224700, 1122

In [ ]:
A=set(l_100)
B=set(l_100_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11224451, 11259908, 11213317, 11219708, 11213187, 11234826, 11219467, 11234827, 11224459, 11253390, 11213195, 11213331, 11219476, 11219731, 11219351, 11253144, 11272986, 11224700, 11229346, 11246758, 11229480, 11235116, 11213236, 11291833, 11229771, 11259851, 11266521, 11266784, 11241318, 11213675, 11213676, 11272944, 11246705, 11229552, 11241329, 11213431, 11246715, 11272956, 11241214, 11260159}
AUB =  {11259908, 11213317, 11219462, 11213319, 11234822, 11234826, 11219467, 11219466, 11213323, 11292173, 11219468, 11234827, 11213331, 11219476, 11234843, 11229735, 11246632, 11250728, 11234857, 11213356, 11241519, 11259953, 11213362, 11219517, 11285056, 11259969, 11219523, 11305028, 11229771, 11229772, 11219532, 11213388, 11246674, 11229780, 11229782, 11233369, 11219547, 11311717, 11213414, 11253355, 11273323, 11246701, 11224687, 11253360, 11246705, 11271281, 11246708, 11213431, 11213434, 11246715, 11224700, 11224702, 11298434, 11253379, 11298437, 11253386, 11253390, 11224720, 1124

In [ ]:
df = pd.read_csv('/content/merged_H04.csv')

In [ ]:
df=df[0:1000]

In [ ]:
df.head()

patent_id                                    patent_abstract  \
0   11221751  An electronic device displays a messaging user...   
1   11225204  An object of the present invention is to provi...   
2   11218777  A method for processing bullet screen, compris...   
3   11218849  A vehicle-to-X communications system for a veh...   
4   11218254  The present disclosure discloses a method and ...   

                                        patent_title  \
0  Devices, methods, and graphical user interface...   
1                 On-vehicle image processing device   
2  Method, device and system for processing bulle...   
3   Vehicle to X communications system for a vehicle   
4  Method and apparatus for transmitting/receivin...   

                                          claim_text cpc_subclass  \
0   A method, comprising: at an electronic device...         H04L   
1   A stereo vision system for vehicle comprising...         H04N   
2   A method for processing bullet screen, compri...         H04N   
3   A vehicle-to-X communications system for a ve...         H04W   
4   A method for transmitting hybrid automatic re...         H04L   

                                  generated_abstract  
0    The patent claim describes a method for disp...  
1    The present patent claim relates to a stereo...  
2    The present disclosure relates to a method f...  
3    The present invention relates to a vehicle-t...  
4    The present patent claim describes a method ...

In [ ]:
df.shape

(1000, 6)

In [ ]:
abstracts = df['patent_abstract'].tolist()
patent_ids = df['patent_id'].tolist()
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

def encode_texts(texts):
    encoded_inputs = tokenizer(texts, padding=True, truncation=True, return_tensors='pt', max_length=512)
    with torch.no_grad():
        outputs = model(**encoded_inputs)
    return outputs.last_hidden_state.mean(dim=1).cpu().numpy()
abstracts_embeddings = encode_texts(abstracts)

similarity_matrix = cosine_similarity(abstracts_embeddings)
similarities = similarity_matrix[0][1:]

In [ ]:
closest_indices_20 = similarities.argsort()[-25:][::-1]
closest_patent_ids_20 = [patent_ids[i+1] for i in closest_indices_20]
l_30=[]
for i, patent_id in enumerate(closest_patent_ids_20):
    l_30.append(patent_id)
print(l_30)
print("=======================================")
closest_indices_10 = similarities.argsort()[-50:][::-1]
closest_patent_ids_10 = [patent_ids[i+1] for i in closest_indices_10]
l_50=[]
for i, patent_id in enumerate(closest_patent_ids_10):
  l_50.append(patent_id)
print(l_50)
print("=======================================")
closest_indices_5 = similarities.argsort()[-75:][::-1]
closest_patent_ids_5 = [patent_ids[i+1] for i in closest_indices_5]
l_75=[]
for i, patent_id in enumerate(closest_patent_ids_5):
    l_75.append(patent_id)
print(l_75)
print("=======================================")
closest_indices_3 = similarities.argsort()[-100:][::-1]
closest_patent_ids_3 = [patent_ids[i+1] for i in closest_indices_3]
l_100=[]
for i, patent_id in enumerate(closest_patent_ids_3):
    l_100.append(patent_id)
print(l_100)

[11231785, 11227356, 11221649, 11231835, 11216147, 11223586, 11223853, 11218531, 11218524, 11216655, 11221599, 11223728, 11218526, 11222030, 11218444, 11216865, 11218547, 11228703, 11222538, 11215416, 11284419, 11216342, 11222726, 11222726, 11227598]
[11231785, 11227356, 11221649, 11231835, 11216147, 11223586, 11223853, 11218531, 11218524, 11216655, 11221599, 11223728, 11218526, 11222030, 11218444, 11216865, 11218547, 11228703, 11222538, 11215416, 11284419, 11216342, 11222726, 11222726, 11227598, 11216555, 11218666, 11277690, 11227485, 11228861, 11218469, 11223741, 11218568, 11223762, 11219852, 11228736, 11218771, 11221892, 11221892, 11218401, 11217343, 11222359, 11218768, 11218365, 11227483, 11233916, 11227025, 11214263, 11218294, 11223548]
[11231785, 11227356, 11221649, 11231835, 11216147, 11223586, 11223853, 11218531, 11218524, 11216655, 11221599, 11223728, 11218526, 11222030, 11218444, 11216865, 11218547, 11228703, 11222538, 11215416, 11284419, 11216342, 11222726, 11222726, 1122759

In [ ]:
abstracts_g = df['generated_abstract'].tolist()
patent_ids = df['patent_id'].tolist()
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

def encode_texts(texts):
    encoded_inputs = tokenizer(texts, padding=True, truncation=True, return_tensors='pt', max_length=512)
    with torch.no_grad():
        outputs = model(**encoded_inputs)
    return outputs.last_hidden_state.mean(dim=1).cpu().numpy()

abstracts_embeddings_g = encode_texts(abstracts_g)

similarity_matrix_g = cosine_similarity(abstracts_embeddings_g)
similarities_g = similarity_matrix_g[0][1:]

In [ ]:
closest_indices_20 = similarities_g.argsort()[-25:][::-1]
closest_patent_ids_20 = [patent_ids[i+1] for i in closest_indices_20]
l_25_g=[]
for i, patent_id in enumerate(closest_patent_ids_20):
  l_25_g.append(patent_id)
print(l_25_g)
print("=======================================")

closest_indices_10 = similarities_g.argsort()[-50:][::-1]
closest_patent_ids_10 = [patent_ids[i+1] for i in closest_indices_10]
l_50_g=[]
for i, patent_id in enumerate(closest_patent_ids_10):
  l_50_g.append(patent_id)
print(l_50_g)
print("=======================================")
closest_indices_5 = similarities_g.argsort()[-75:][::-1]
closest_patent_ids_5 = [patent_ids[i+1] for i in closest_indices_5]
l_75_g=[]
for i, patent_id in enumerate(closest_patent_ids_5):
  l_75_g.append(patent_id)
print(l_75_g)
print("=======================================")
closest_indices_3 = similarities_g.argsort()[-100:][::-1]
closest_patent_ids_3 = [patent_ids[i+1] for i in closest_indices_3]
l_100_g=[]
for i, patent_id in enumerate(closest_patent_ids_3):
  l_100_g.append(patent_id)
print(l_100_g)

[11221649, 11216585, 11223586, 11222030, 11218433, 11218666, 11223728, 11227598, 11223771, 11240469, 11228702, 11284334, 11222538, 11218412, 11228703, 11228617, 11221646, 11233740, 11219065, 11216147, 11224028, 11218777, 11275552, 11218531, 11216547]
[11221649, 11216585, 11223586, 11222030, 11218433, 11218666, 11223728, 11227598, 11223771, 11240469, 11228702, 11284334, 11222538, 11218412, 11228703, 11228617, 11221646, 11233740, 11219065, 11216147, 11224028, 11218777, 11275552, 11218531, 11216547, 11218782, 11246111, 11228888, 11284283, 11218524, 11215416, 11218796, 11216342, 11220238, 11223742, 11222061, 11224040, 11224783, 11223729, 11302165, 11228563, 11216846, 11222400, 11223521, 11218444, 11218516, 11251926, 11221892, 11217027, 11284214]
[11221649, 11216585, 11223586, 11222030, 11218433, 11218666, 11223728, 11227598, 11223771, 11240469, 11228702, 11284334, 11222538, 11218412, 11228703, 11228617, 11221646, 11233740, 11219065, 11216147, 11224028, 11218777, 11275552, 11218531, 1121654

In [ ]:
A =set(l_30)
B = set(l_25_g)
# Intersaction and Union of two sets can also be done using & and | operators.
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11223586, 11218531, 11222538, 11222030, 11227598, 11223728, 11221649, 11216147, 11228703}
AUB =  {11218433, 11222538, 11218444, 11222030, 11216655, 11221646, 11221649, 11216147, 11240469, 11228702, 11228703, 11275552, 11223586, 11216547, 11231785, 11223853, 11223728, 11215416, 11284419, 11222726, 11216585, 11228617, 11233740, 11227598, 11216342, 11218777, 11231835, 11227356, 11218524, 11218526, 11221599, 11223771, 11216865, 11224028, 11218531, 11218666, 11218412, 11284334, 11218547, 11219065}
J(A,B) =  0.225


In [ ]:
A=set(l_50)
B=set(l_50_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))


AnB =  {11223586, 11218531, 11221892, 11222538, 11218666, 11218444, 11222030, 11227598, 11223728, 11221649, 11216147, 11216342, 11215416, 11218524, 11228703}
AUB =  {11218433, 11222538, 11218444, 11222030, 11216655, 11220238, 11216147, 11240469, 11302165, 11228702, 11228703, 11275552, 11246111, 11223586, 11218469, 11231785, 11223853, 11222061, 11215416, 11284283, 11228736, 11218768, 11218771, 11218516, 11222359, 11218777, 11231835, 11218524, 11227485, 11218526, 11221599, 11227483, 11218782, 11218531, 11218796, 11284334, 11218547, 11218294, 11219065, 11277690, 11233916, 11222400, 11217027, 11221892, 11218568, 11219852, 11221646, 11221649, 11227025, 11228563, 11216547, 11216555, 11223728, 11223729, 11214263, 11228861, 11223741, 11217343, 11218365, 11223742, 11284419, 11222726, 11216585, 11228617, 11233740, 11227598, 11224783, 11216846, 11223762, 11216342, 11251926, 11228888, 11223771, 11227356, 11224028, 11216865, 11218401, 11223521, 11224040, 11218666, 11218412, 11284214, 11223548}
J(A,

In [ ]:
A=set(l_75)
B=set(l_75_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11221892, 11218568, 11222538, 11233931, 11218444, 11222030, 11221646, 11221649, 11216147, 11228703, 11223586, 11218851, 11223853, 11223728, 11215416, 11284283, 11228736, 11233916, 11216585, 11233740, 11227598, 11216342, 11218524, 11218531, 11218666, 11228796}
AUB =  {11223553, 11218433, 11222538, 11218444, 11222030, 11216655, 11220238, 11216147, 11240469, 11228438, 11224087, 11302165, 11295002, 11228702, 11228703, 11275552, 11246111, 11223586, 11218469, 11284263, 11234088, 11231785, 11228716, 11223853, 11222061, 11234101, 11221814, 11215416, 11221818, 11284283, 11222330, 11228736, 11218753, 11228739, 11217220, 11219013, 11218768, 11216977, 11218771, 11218516, 11222359, 11218777, 11231835, 11218524, 11227485, 11218526, 11221599, 11227483, 11218782, 11218531, 11218796, 11284334, 11215983, 11218547, 11218294, 11215991, 11222646, 11219065, 11277690, 11240060, 11228796, 11233916, 11222400, 11217027, 11221892, 11218568, 11233931, 11219852, 11221646, 11221649, 11227025, 11228563, 1121

In [ ]:
A=set(l_100)
B=set(l_100_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11223553, 11221892, 11218568, 11222538, 11233931, 11218444, 11222030, 11221646, 11213455, 11221649, 11216147, 11244564, 11240469, 11228438, 11246111, 11228703, 11223586, 11218851, 11223853, 11223728, 11215416, 11284283, 11228736, 11233916, 11216585, 11233740, 11227598, 11224783, 11218768, 11216342, 11218524, 11218401, 11218531, 11218666, 11218412, 11284334, 11218547, 11227130, 11228796}
AUB =  {11223553, 11218433, 11222538, 11224075, 11218444, 11222030, 11218449, 11244564, 11240469, 11228693, 11224087, 11228702, 11228703, 11246111, 11223586, 11218469, 11231785, 11222058, 11228716, 11222061, 11218478, 11215416, 11228736, 11228739, 11219013, 11228748, 11216977, 11218516, 11219032, 11231835, 11218524, 11218526, 11221599, 11218531, 11218536, 11215980, 11215983, 11218547, 11219059, 11218548, 11222646, 11215991, 11219065, 11240060, 11228796, 11233916, 11217027, 11218568, 11233931, 11221646, 11213455, 11221649, 11212953, 11233950, 11216547, 11240616, 11216555, 11223725, 11223728, 1122

In [ ]:
df = pd.read_csv('/content/merged_G06.csv')

In [ ]:
df=df[0:1000]

In [ ]:
df.head()

patent_id                                    patent_abstract  \
0   11232283  Facial recognition system that compares narrow...   
1   11216575  In one embodiment, data at rest is securely st...   
2   11216801  A voice controlled (VC) system for onboarding ...   
3   11229489  Systems and methods may be used to perform rob...   
4   11238654  Embodiments of the specification provide offli...   

                                        patent_title  \
0  Facial recognition system that compares narrow...   
1  Enhanced securing and secured processing of da...   
2  Voice controlled systems and methods for onboa...   
3         Soft tissue balancing in articular surgery   
4        Offline shopping guide method and apparatus   

                                          claim_text cpc_subclass  \
0   A facial recognition system that compares nar...         G06V   
1   An apparatus, comprising: a data safe perform...         G06F   
2   A voice controlled (VC) system for onboarding...         G06Q   
3   An intraoperative surgical system comprising:...         G06T   
4   An offline shopping guide method, comprising:...         G06V   

                                  generated_abstract  
0    The present invention relates to a facial re...  
1    The present patent claim relates to an appar...  
2    The present invention relates to a voice con...  
3    The present invention relates to an intraope...  
4    The present patent describes an offline shop...

In [ ]:
df.shape

(1000, 6)

In [ ]:
abstracts = df['patent_abstract'].tolist()
patent_ids = df['patent_id'].tolist()
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

def encode_texts(texts):
    encoded_inputs = tokenizer(texts, padding=True, truncation=True, return_tensors='pt', max_length=512)
    with torch.no_grad():
        outputs = model(**encoded_inputs)
    return outputs.last_hidden_state.mean(dim=1).cpu().numpy()
abstracts_embeddings = encode_texts(abstracts)

similarity_matrix = cosine_similarity(abstracts_embeddings)
similarities = similarity_matrix[0][1:]

In [ ]:
closest_indices_20 = similarities.argsort()[-25:][::-1]
closest_patent_ids_20 = [patent_ids[i+1] for i in closest_indices_20]
l_30=[]
for i, patent_id in enumerate(closest_patent_ids_20):
    l_30.append(patent_id)
print(l_30)
print("=======================================")
closest_indices_10 = similarities.argsort()[-50:][::-1]
closest_patent_ids_10 = [patent_ids[i+1] for i in closest_indices_10]
l_50=[]
for i, patent_id in enumerate(closest_patent_ids_10):
  l_50.append(patent_id)
print(l_50)
print("=======================================")
closest_indices_5 = similarities.argsort()[-75:][::-1]
closest_patent_ids_5 = [patent_ids[i+1] for i in closest_indices_5]
l_75=[]
for i, patent_id in enumerate(closest_patent_ids_5):
    l_75.append(patent_id)
print(l_75)
print("=======================================")
closest_indices_3 = similarities.argsort()[-100:][::-1]
closest_patent_ids_3 = [patent_ids[i+1] for i in closest_indices_3]
l_100=[]
for i, patent_id in enumerate(closest_patent_ids_3):
    l_100.append(patent_id)
print(l_100)

[11222466, 11227458, 11227458, 11216988, 11216669, 11226193, 11227405, 11227152, 11215837, 11224483, 11267276, 11235613, 11321597, 11218668, 11238272, 11218648, 11218648, 11238615, 11229496, 11238580, 11216681, 11216681, 11295104, 11250556, 11232285]
[11222466, 11227458, 11227458, 11216988, 11216669, 11226193, 11227405, 11227152, 11215837, 11224483, 11267276, 11235613, 11321597, 11218668, 11238272, 11218648, 11218648, 11238615, 11229496, 11238580, 11216681, 11216681, 11295104, 11250556, 11232285, 11216890, 11243172, 11232293, 11238274, 11218629, 11215845, 11222420, 11227424, 11232595, 11232594, 11216998, 11237107, 11222217, 11222203, 11222452, 11227156, 11243117, 11224399, 11216680, 11224799, 11216784, 11216922, 11227161, 11232289, 11240460]
[11222466, 11227458, 11227458, 11216988, 11216669, 11226193, 11227405, 11227152, 11215837, 11224483, 11267276, 11235613, 11321597, 11218668, 11238272, 11218648, 11218648, 11238615, 11229496, 11238580, 11216681, 11216681, 11295104, 11250556, 1123228

In [ ]:
abstracts_g = df['generated_abstract'].tolist()
patent_ids = df['patent_id'].tolist()
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertModel.from_pretrained('bert-base-uncased')

def encode_texts(texts):
    encoded_inputs = tokenizer(texts, padding=True, truncation=True, return_tensors='pt', max_length=512)
    with torch.no_grad():
        outputs = model(**encoded_inputs)
    return outputs.last_hidden_state.mean(dim=1).cpu().numpy()

abstracts_embeddings_g = encode_texts(abstracts_g)

similarity_matrix_g = cosine_similarity(abstracts_embeddings_g)
similarities_g = similarity_matrix_g[0][1:]

In [ ]:
closest_indices_20 = similarities_g.argsort()[-25:][::-1]
closest_patent_ids_20 = [patent_ids[i+1] for i in closest_indices_20]
l_25_g=[]
for i, patent_id in enumerate(closest_patent_ids_20):
  l_25_g.append(patent_id)
print(l_25_g)
print("=======================================")

closest_indices_10 = similarities_g.argsort()[-50:][::-1]
closest_patent_ids_10 = [patent_ids[i+1] for i in closest_indices_10]
l_50_g=[]
for i, patent_id in enumerate(closest_patent_ids_10):
  l_50_g.append(patent_id)
print(l_50_g)
print("=======================================")
closest_indices_5 = similarities_g.argsort()[-75:][::-1]
closest_patent_ids_5 = [patent_ids[i+1] for i in closest_indices_5]
l_75_g=[]
for i, patent_id in enumerate(closest_patent_ids_5):
  l_75_g.append(patent_id)
print(l_75_g)
print("=======================================")
closest_indices_3 = similarities_g.argsort()[-100:][::-1]
closest_patent_ids_3 = [patent_ids[i+1] for i in closest_indices_3]
l_100_g=[]
for i, patent_id in enumerate(closest_patent_ids_3):
  l_100_g.append(patent_id)
print(l_100_g)

[11216983, 11218629, 11232289, 11216949, 11238230, 11216894, 11227458, 11227141, 11238272, 11216681, 11222358, 11222439, 11218429, 11232306, 11250226, 11222452, 11227438, 11232287, 11216927, 11216922, 11232594, 11216697, 11218429, 11216686, 11223571]
[11216983, 11218629, 11232289, 11216949, 11238230, 11216894, 11227458, 11227141, 11238272, 11216681, 11222358, 11222439, 11218429, 11232306, 11250226, 11222452, 11227438, 11232287, 11216927, 11216922, 11232594, 11216697, 11218429, 11216686, 11223571, 11227458, 11222261, 11222494, 11213201, 11334326, 11216988, 11236453, 11232512, 11216833, 11218668, 11238274, 11328482, 11227191, 11222365, 11227407, 11238839, 11216681, 11238580, 11226697, 11216499, 11216933, 11222641, 11334738, 11232315, 11216987]
[11216983, 11218629, 11232289, 11216949, 11238230, 11216894, 11227458, 11227141, 11238272, 11216681, 11222358, 11222439, 11218429, 11232306, 11250226, 11222452, 11227438, 11232287, 11216927, 11216922, 11232594, 11216697, 11218429, 11216686, 1122357

In [ ]:
A =set(l_30)
B = set(l_25_g)
# Intersaction and Union of two sets can also be done using & and | operators.
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11238272, 11216681, 11227458}
AUB =  {11238272, 11295104, 11227141, 11227405, 11227152, 11223571, 11216922, 11235613, 11216669, 11232285, 11216927, 11232287, 11232289, 11224483, 11222439, 11216681, 11227438, 11216686, 11232306, 11250226, 11238580, 11222452, 11216949, 11229496, 11216697, 11222466, 11227458, 11218629, 11267276, 11226193, 11232594, 11238230, 11238615, 11218648, 11216983, 11222358, 11216988, 11215837, 11218429, 11218668, 11250556, 11321597, 11216894}
J(A,B) =  0.06976744186046512


In [ ]:
A=set(l_50)
B=set(l_50_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))


AnB =  {11238272, 11232289, 11238274, 11227458, 11218629, 11216681, 11218668, 11232594, 11222452, 11238580, 11216922, 11216988}
AUB =  {11232512, 11227141, 11240460, 11227405, 11227407, 11227152, 11223571, 11227156, 11227161, 11216922, 11235613, 11216669, 11232285, 11227424, 11232289, 11216927, 11232287, 11232293, 11216933, 11216680, 11216681, 11227438, 11216686, 11232306, 11250226, 11216949, 11227191, 11229496, 11216697, 11232315, 11227458, 11226697, 11224399, 11226193, 11232594, 11232595, 11334738, 11238230, 11216983, 11222358, 11216987, 11216988, 11222365, 11218429, 11236453, 11216998, 11243117, 11222641, 11216499, 11250556, 11238272, 11295104, 11238274, 11216784, 11213201, 11222420, 11224483, 11243172, 11222439, 11238580, 11222452, 11334326, 11238839, 11222203, 11216833, 11222466, 11218629, 11222217, 11267276, 11238615, 11218648, 11215837, 11222494, 11224799, 11328482, 11215845, 11218668, 11237107, 11222261, 11216890, 11321597, 11216894}
J(A,B) =  0.14634146341463414


In [ ]:
A=set(l_75)
B=set(l_75_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11238272, 11238274, 11227141, 11227407, 11227152, 11227153, 11216922, 11232289, 11224483, 11216681, 11222452, 11216949, 11238580, 11227458, 11218629, 11232594, 11216988, 11218668, 11216894}
AUB =  {11232512, 11216640, 11227141, 11227142, 11227146, 11240460, 11227405, 11227407, 11227152, 11227153, 11216655, 11223571, 11227156, 11301655, 11227161, 11216922, 11235613, 11216669, 11232285, 11227424, 11232289, 11227170, 11216926, 11216927, 11232293, 11232287, 11216933, 11216680, 11216681, 11232549, 11232296, 11227438, 11216686, 11216690, 11232306, 11250226, 11216949, 11227187, 11227191, 11229496, 11216697, 11217208, 11232315, 11238973, 11216702, 11226687, 11232575, 11227458, 11226697, 11234634, 11234633, 11228747, 11224399, 11226193, 11232594, 11232595, 11334738, 11238230, 11216983, 11222358, 11216987, 11216988, 11222365, 11223645, 11218429, 11222627, 11236453, 11216998, 11216999, 11243117, 11222641, 11216499, 11244153, 11250556, 11215997, 11238272, 11295104, 11238274, 11275917, 1121

In [ ]:
A=set(l_100)
B=set(l_100_g)
C = A.intersection(B)
D = A.union(B)
print('AnB = ', C)
print('AUB = ', D)
print('J(A,B) = ', float(len(C))/float(len(D)))

AnB =  {11238272, 11238274, 11227141, 11227407, 11227152, 11227153, 11227156, 11216922, 11232289, 11224483, 11222439, 11216681, 11238580, 11216949, 11222452, 11217208, 11232315, 11227458, 11218629, 11232594, 11238615, 11216988, 11222494, 11224799, 11216999, 11218668, 11227384, 11216894}
AUB =  {11218433, 11227141, 11227142, 11227146, 11240460, 11227151, 11227152, 11227153, 11216912, 11223571, 11227156, 11333654, 11227161, 11216922, 11232285, 11216926, 11216927, 11232287, 11232289, 11227170, 11232293, 11216933, 11232296, 11216426, 11232306, 11231283, 11250226, 11216949, 11227187, 11227191, 11232315, 11238973, 11226687, 11226697, 11228747, 11334733, 11226193, 11334738, 11243607, 11216983, 11216987, 11216988, 11223645, 11222627, 11222628, 11236453, 11216998, 11216999, 11243117, 11222641, 11217011, 11216499, 11244153, 11215997, 11229824, 11228810, 11275917, 11328143, 11233944, 11227291, 11243172, 11222191, 11238576, 11238580, 11244212, 11229364, 11334326, 11222203, 11222205, 11218629, 1121